File: ciphex-predictions/src/features/abacus-index/utils/barBuilder.ts
/**
 * Bar Builder Utility
 *
 * Constructs 1-minute OHLCV bars from trade streams.
 * Handles partial bars, bar completion, and gap detection.
 */

import { Trade, Bar, VenueId, AssetId, MarketType } from '../types';
import { BAR_INTERVAL_SECONDS, MAX_TRADE_BUFFER_SIZE } from '../constants';

// =============================================================================
// Types
// =============================================================================

export interface BarBuilderState {
  /** Current partial bar being built */
  currentBar: Bar | null;
  /** Completed bars (most recent last) */
  completedBars: Bar[];
  /** Trades buffered for current bar */
  tradeBuffer: Trade[];
  /** Last bar close time (for gap detection) */
  lastBarTime: number | null;
  /** Detected gaps (missing bar times) */
  gaps: number[];
}

export interface BarBuilderConfig {
  venue: VenueId;
  asset: AssetId;
  marketType: MarketType;
  /** Maximum completed bars to retain */
  maxBars?: number;
}

// =============================================================================
// Core Functions
// =============================================================================

/**
 * Create initial bar builder state
 */
export function createBarBuilderState(): BarBuilderState {
  return {
    currentBar: null,
    completedBars: [],
    tradeBuffer: [],
    lastBarTime: null,
    gaps: [],
  };
}

/**
 * Get the bar start time for a given timestamp
 * Floors to the nearest minute boundary
 */
export function getBarTime(timestampMs: number): number {
  const seconds = Math.floor(timestampMs / 1000);
  return Math.floor(seconds / BAR_INTERVAL_SECONDS) * BAR_INTERVAL_SECONDS;
}

/**
 * Process a trade and update bar state
 *
 * @param state - Current bar builder state
 * @param trade - Incoming trade
 * @param config - Bar builder configuration
 * @returns Updated state
 */
export function processTrade(
  state: BarBuilderState,
  trade: Trade,
  config: BarBuilderConfig
): BarBuilderState {
  const barTime = getBarTime(trade.timestamp);
  const { venue, asset, marketType, maxBars = 1000 } = config;

  // Clone state for immutability
  let newState: BarBuilderState = {
    ...state,
    tradeBuffer: [...state.tradeBuffer],
    completedBars: [...state.completedBars],
    gaps: [...state.gaps],
  };

  // Check if this trade belongs to a new bar
  if (newState.currentBar && barTime > newState.currentBar.time) {
    // Complete the current bar
    const completedBar: Bar = {
      ...newState.currentBar,
      isPartial: false,
    };
    newState.completedBars.push(completedBar);

    // Detect gaps (missing bars)
    if (newState.lastBarTime !== null) {
      const expectedBars = (barTime - newState.lastBarTime) / BAR_INTERVAL_SECONDS;
      if (expectedBars > 1) {
        // There are missing bars
        for (let i = 1; i < expectedBars; i++) {
          const missingTime = newState.lastBarTime + i * BAR_INTERVAL_SECONDS;
          newState.gaps.push(missingTime);
        }
      }
    }

    newState.lastBarTime = completedBar.time;
    newState.currentBar = null;
    newState.tradeBuffer = [];

    // Trim old bars if needed
    if (newState.completedBars.length > maxBars) {
      newState.completedBars = newState.completedBars.slice(-maxBars);
    }
  }

  // Initialize new bar if needed
  if (!newState.currentBar || barTime > newState.currentBar.time) {
    newState.currentBar = {
      time: barTime,
      open: trade.price,
      high: trade.price,
      low: trade.price,
      close: trade.price,
      volume: trade.quantity,
      tradeCount: 1,
      venue,
      asset,
      marketType,
      isPartial: true,
    };
    newState.tradeBuffer = [trade];
  } else {
    // Update existing bar
    newState.currentBar = {
      ...newState.currentBar,
      high: Math.max(newState.currentBar.high, trade.price),
      low: Math.min(newState.currentBar.low, trade.price),
      close: trade.price,
      volume: newState.currentBar.volume + trade.quantity,
      tradeCount: newState.currentBar.tradeCount + 1,
    };

    // Buffer trade (with size limit)
    if (newState.tradeBuffer.length < MAX_TRADE_BUFFER_SIZE) {
      newState.tradeBuffer.push(trade);
    }
  }

  return newState;
}

/**
 * Force-complete the current bar (e.g., on reconnect)
 * Marks the bar as potentially incomplete
 */
export function forceCompleteBar(state: BarBuilderState): BarBuilderState {
  if (!state.currentBar) {
    return state;
  }

  const completedBar: Bar = {
    ...state.currentBar,
    isPartial: false, // Mark as complete even though it may be incomplete
  };

  return {
    ...state,
    currentBar: null,
    completedBars: [...state.completedBars, completedBar],
    tradeBuffer: [],
    lastBarTime: completedBar.time,
  };
}

/**
 * Discard the current partial bar (e.g., on reconnect when we can't trust it)
 *
 * Gap recording logic:
 * - Only record a gap if we've completed at least one bar (lastBarTime !== null)
 * - This prevents spurious gap counts from cold-start scenarios:
 *   - React strict mode double-mount
 *   - Initial WebSocket connection before first bar completes
 * - The gap detection in processTrade() will catch any missed bars on reconnect
 */
export function discardPartialBar(state: BarBuilderState): BarBuilderState {
  if (!state.currentBar) {
    return state;
  }

  // Only record gap if we have established continuity (completed at least one bar)
  // Cold-start discards should not count as gaps
  const newGaps = [...state.gaps];
  if (state.lastBarTime !== null) {
    newGaps.push(state.currentBar.time);
  }

  return {
    ...state,
    currentBar: null,
    tradeBuffer: [],
    gaps: newGaps,
  };
}

/**
 * Get all bars (completed + current partial)
 */
export function getAllBars(state: BarBuilderState): Bar[] {
  if (state.currentBar) {
    return [...state.completedBars, state.currentBar];
  }
  return state.completedBars;
}

/**
 * Get the current close price
 */
export function getCurrentPrice(state: BarBuilderState): number | null {
  return state.currentBar?.close ?? state.completedBars[state.completedBars.length - 1]?.close ?? null;
}

// =============================================================================
// Batch Processing
// =============================================================================

/**
 * Process multiple trades at once (for initial data load)
 */
export function processTradesBatch(
  state: BarBuilderState,
  trades: Trade[],
  config: BarBuilderConfig
): BarBuilderState {
  let currentState = state;
  for (const trade of trades) {
    currentState = processTrade(currentState, trade, config);
  }
  return currentState;
}

// =============================================================================
// Bar Merging (for composites)
// =============================================================================

/**
 * Align bars from multiple sources by time
 * Returns bars that exist in all sources at each time
 *
 * Optimized to O(N) using Map lookups instead of O(N²) with .find()
 */
export function alignBars(barSets: Bar[][]): Map<number, Bar[]> {
  const aligned = new Map<number, Bar[]>();

  // Build a lookup Map for each bar set: time -> Bar
  // This gives O(1) lookup per time instead of O(N) with .find()
  const barMaps: Map<number, Bar>[] = barSets.map((bars) => {
    const map = new Map<number, Bar>();
    for (const bar of bars) {
      map.set(bar.time, bar);
    }
    return map;
  });

  // Collect all unique times
  const allTimes = new Set<number>();
  for (const bars of barSets) {
    for (const bar of bars) {
      allTimes.add(bar.time);
    }
  }

  // For each time, collect bars from each source using O(1) Map lookup
  for (const time of allTimes) {
    const barsAtTime: Bar[] = [];
    for (const barMap of barMaps) {
      const bar = barMap.get(time);
      if (bar) {
        barsAtTime.push(bar);
      }
    }
    aligned.set(time, barsAtTime);
  }

  return aligned;
}

File: ciphex-predictions/services/abacus_indexer/core/bar_builder.py
"""
Abacus Indexer Bar Builder

Constructs 1-minute OHLCV bars from incoming trades.
Handles bar accumulation, completion detection, and history management.
"""

from collections import deque
from dataclasses import dataclass, field
from typing import Callable, Optional

from .constants import BAR_INTERVAL_SECONDS, MAX_BARS_PER_VENUE, MAX_TRADE_BUFFER_SIZE
from .types import AssetId, Bar, MarketType, TakerSide, Trade, VenueId


def floor_to_minute(timestamp_ms: int) -> int:
    """Floor a timestamp (ms) to the start of its minute (unix seconds)."""
    return (timestamp_ms // 1000 // BAR_INTERVAL_SECONDS) * BAR_INTERVAL_SECONDS


@dataclass
class BarAccumulator:
    """
    Accumulates trades into a forming bar.

    This is a mutable state object that collects trades for the current minute
    and produces a completed Bar when the minute boundary is crossed.

    Buy/Sell volume separation follows the normalized taker-side semantics:
    - buy_volume/buy_count: Taker-initiated buys (aggressive buys)
    - sell_volume/sell_count: Taker-initiated sells (aggressive sells)
    """

    bar_time: int  # Unix seconds (start of bar)
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: Optional[float] = None
    volume: float = 0.0
    trade_count: int = 0

    # Buy/Sell volume separation (taker-initiated)
    buy_volume: float = 0.0
    sell_volume: float = 0.0
    buy_count: int = 0
    sell_count: int = 0

    venue: Optional[VenueId] = None
    asset: Optional[AssetId] = None
    market_type: Optional[MarketType] = None

    def add_trade(self, trade: Trade) -> None:
        """
        Add a trade to the accumulator.

        Assumes trade belongs to this bar's minute (caller should verify).
        Uses trade.taker_side for normalized buy/sell classification.
        """
        price = trade.price
        quantity = trade.quantity

        # First trade sets all values
        if self.open is None:
            self.open = price
            self.high = price
            self.low = price
            self.close = price
            self.venue = trade.venue
            self.asset = trade.asset
            self.market_type = trade.market_type
        else:
            # Update OHLC
            if price > self.high:  # type: ignore
                self.high = price
            if price < self.low:  # type: ignore
                self.low = price
            self.close = price

        # Accumulate total volume and count
        self.volume += quantity
        self.trade_count += 1

        # Accumulate buy/sell volume using normalized taker_side
        if trade.taker_side == TakerSide.BUY:
            self.buy_volume += quantity
            self.buy_count += 1
        else:
            self.sell_volume += quantity
            self.sell_count += 1

    def to_bar(self, is_partial: bool = False) -> Optional[Bar]:
        """
        Convert accumulator to a Bar.

        Returns None if no trades have been accumulated.
        """
        if self.open is None or self.venue is None or self.asset is None or self.market_type is None:
            return None

        return Bar(
            time=self.bar_time,
            open=self.open,
            high=self.high,  # type: ignore
            low=self.low,  # type: ignore
            close=self.close,  # type: ignore
            volume=self.volume,
            trade_count=self.trade_count,
            buy_volume=self.buy_volume,
            sell_volume=self.sell_volume,
            buy_count=self.buy_count,
            sell_count=self.sell_count,
            venue=self.venue,
            asset=self.asset,
            market_type=self.market_type,
            is_partial=is_partial,
        )

    def reset(self, new_bar_time: int) -> None:
        """Reset accumulator for a new bar."""
        self.bar_time = new_bar_time
        self.open = None
        self.high = None
        self.low = None
        self.close = None
        self.volume = 0.0
        self.trade_count = 0
        self.buy_volume = 0.0
        self.sell_volume = 0.0
        self.buy_count = 0
        self.sell_count = 0


@dataclass
class BarBuilder:
    """
    Builds 1-minute OHLCV bars from a stream of trades.

    Features:
    - Automatic bar completion on minute boundary crossing
    - Partial bar support (current forming bar)
    - History management with ring buffer (MAX_BARS_PER_VENUE)
    - Callback on bar completion

    Usage:
        builder = BarBuilder(
            venue=VenueId.BINANCE,
            asset=AssetId.BTC,
            market_type=MarketType.SPOT,
            on_bar_complete=handle_completed_bar,
        )
        builder.add_trade(trade)
        partial_bar = builder.get_partial_bar()
        history = builder.get_bars()
    """

    venue: VenueId
    asset: AssetId
    market_type: MarketType
    on_bar_complete: Optional[Callable[[Bar], None]] = None

    # Internal state
    _accumulator: Optional[BarAccumulator] = field(default=None, init=False)
    _completed_bars: deque[Bar] = field(default_factory=lambda: deque(maxlen=MAX_BARS_PER_VENUE), init=False)
    _trade_count: int = field(default=0, init=False)
    _last_trade_time: Optional[int] = field(default=None, init=False)

    def add_trade(self, trade: Trade) -> Optional[Bar]:
        """
        Add a trade to the builder.

        Args:
            trade: Incoming trade

        Returns:
            Completed bar if a bar was closed, None otherwise
        """
        trade_bar_time = floor_to_minute(trade.timestamp)

        # Initialize accumulator if needed
        if self._accumulator is None:
            self._accumulator = BarAccumulator(bar_time=trade_bar_time)

        completed_bar: Optional[Bar] = None

        # Check if trade belongs to a new bar - MUST happen before trade count check
        # so that _trade_count gets reset even if we were at the limit
        if trade_bar_time > self._accumulator.bar_time:
            # Complete the current bar
            completed_bar = self._accumulator.to_bar(is_partial=False)
            if completed_bar:
                self._completed_bars.append(completed_bar)
                if self.on_bar_complete:
                    self.on_bar_complete(completed_bar)

            # Fill any gap bars (no trades for entire minutes)
            # We don't create synthetic bars - gaps are handled at composite level
            self._accumulator.reset(trade_bar_time)
            self._trade_count = 0

        # Safety check: limit trades per minute (AFTER bar time check to ensure reset)
        if self._trade_count >= MAX_TRADE_BUFFER_SIZE:
            return completed_bar  # Return any completed bar but don't process this trade

        # Add trade to current accumulator
        self._accumulator.add_trade(trade)
        self._trade_count += 1
        self._last_trade_time = trade.timestamp

        return completed_bar

    def get_partial_bar(self) -> Optional[Bar]:
        """
        Get the current forming bar (partial).

        Returns None if no trades received yet.
        """
        if self._accumulator is None:
            return None
        return self._accumulator.to_bar(is_partial=True)

    def get_bars(self, limit: Optional[int] = None) -> list[Bar]:
        """
        Get completed bars (oldest first).

        Args:
            limit: Optional limit on number of bars to return

        Returns:
            List of completed bars
        """
        bars = list(self._completed_bars)
        if limit:
            return bars[-limit:]
        return bars

    def get_latest_bar(self) -> Optional[Bar]:
        """Get the most recently completed bar."""
        if not self._completed_bars:
            return None
        return self._completed_bars[-1]

    def get_current_price(self) -> Optional[float]:
        """Get the current price (close of partial bar)."""
        if self._accumulator is None:
            return None
        return self._accumulator.close

    def get_last_trade_time(self) -> Optional[int]:
        """Get the timestamp of the last trade received (ms)."""
        return self._last_trade_time

    @property
    def bar_count(self) -> int:
        """Number of completed bars in history."""
        return len(self._completed_bars)


def merge_bars(bars: list[Bar]) -> Optional[Bar]:
    """
    Merge multiple bars into a single bar.

    Used for combining bars from different sources or
    creating aggregated timeframes.

    Args:
        bars: List of bars to merge (must be from same time, venue, asset)

    Returns:
        Merged bar or None if no bars provided
    """
    if not bars:
        return None

    # Use first bar as template
    first = bars[0]

    open_price = first.open
    high_price = first.high
    low_price = first.low
    close_price = first.close
    total_volume = first.volume
    total_trades = first.trade_count
    total_buy_volume = first.buy_volume
    total_sell_volume = first.sell_volume
    total_buy_count = first.buy_count
    total_sell_count = first.sell_count

    for bar in bars[1:]:
        if bar.high > high_price:
            high_price = bar.high
        if bar.low < low_price:
            low_price = bar.low
        close_price = bar.close
        total_volume += bar.volume
        total_trades += bar.trade_count
        total_buy_volume += bar.buy_volume
        total_sell_volume += bar.sell_volume
        total_buy_count += bar.buy_count
        total_sell_count += bar.sell_count

    return Bar(
        time=first.time,
        open=open_price,
        high=high_price,
        low=low_price,
        close=close_price,
        volume=total_volume,
        trade_count=total_trades,
        buy_volume=total_buy_volume,
        sell_volume=total_sell_volume,
        buy_count=total_buy_count,
        sell_count=total_sell_count,
        venue=first.venue,
        asset=first.asset,
        market_type=first.market_type,
        is_partial=any(b.is_partial for b in bars),
    )
